pxtlib/browserutils.ts

```diff
@@ -527,4 +527,330 @@ namespace pxt.BrowserUtils {
+                    this.quotaExceededHandler();
+            }
+        }
+
+        private getObjectStore(name: string, mode: "readonly" | "readwrite" = "readonly"): IDBObjectStore {
+            this.throwIfNotOpened();
+            const transaction = this._db.transaction([name], mode);
+            return transaction.objectStore(name);
+        }
+
+        static deleteDatabaseAsync(name: string): Promise<void> {
+            return new Promise((resolve, reject) => {
+                const idbFactory: IDBFactory = window.indexedDB || (<any>window).mozIndexedDB || (<any>window).webkitIndexedDB || (<any>window).msIndexedDB;
+                const request = idbFactory.deleteDatabase(name);
+                request.onsuccess = () => resolve();
+                request.onerror = () => reject(request.error);
+            });
+        }
+
+        public openAsync(): Promise<void> {
+            return new Promise((resolve, reject) => {
+                const idbFactory: IDBFactory = window.indexedDB || (<any>window).mozIndexedDB || (<any>window).webkitIndexedDB || (<any>window).msIndexedDB;
+                const request = idbFactory.open(this.name, this.version);
+                request.onsuccess = () => {
+                    this._db = request.result;
+                    resolve();
+                };
+                request.onerror = () => this.errorHandler(request.error, "open", reject);
+                request.onupgradeneeded = (ev) => this.upgradeHandler(ev, request);
+            });
+        }
+
+        public getAsync<T>(storeName: string, id: string): Promise<T> {
+            return new Promise((resolve, reject) => {
+                const store = this.getObjectStore(storeName);
+                const request = store.get(id);
+                request.onsuccess = () => resolve(request.result as T);
+                request.onerror = () => this.errorHandler(request.error, "get", reject);
+            });
+        }
+
+        public getAllAsync<T>(storeName: string): Promise<T[]> {
+            return new Promise((resolve, reject) => {
+                const store = this.getObjectStore(storeName);
+                const cursor = store.openCursor();
+                const data: T[] = [];
+
+                cursor.onsuccess = () => {
+                    if (cursor.result) {
+                        data.push(cursor.result.value);
+                        cursor.result.continue();
+                    } else {
+                        resolve(data);
+                    }
+                };
+                cursor.onerror = () => this.errorHandler(cursor.error, "getAll", reject);
+            });
+        }
+
+        public setAsync(storeName: string, data: any): Promise<void> {
+            return new Promise((resolve, reject) => {
+                const store = this.getObjectStore(storeName, "readwrite");
+                let request: IDBRequest;
+
+                if (typeof data.id !== "undefined" && data.id !== null) {
```

Comment:
data.id == null

Don't need to use typeof here

Reply:
Just moving that code.